README.md
Kredit integration/staging tree
================================

http://www.Kredit.org

Copyright (c) 2009-2013 Kredit Developers

What is Kredit?
----------------

Kredit is an experimental new digital currency that enables instant payments to
anyone, anywhere in the world. Kredit uses peer-to-peer technology to operate
with no central authority: managing transactions and issuing money are carried
out collectively by the network. Kredit is also the name of the open source
software which enables the use of this currency.

For more information, as well as an immediately useable, binary version of
the Kredit client software, see http://www.Kredit.org.

License
-------

Kredit is released under the terms of the MIT license. See `COPYING` for more
information or see http://opensource.org/licenses/MIT.

Development process
-------------------

Developers work in their own trees, then submit pull requests when they think
their feature or bug fix is ready.

If it is a simple/trivial/non-controversial change, then one of the Kredit
development team members simply pulls it.

If it is a *more complicated or potentially controversial* change, then the patch
submitter will be asked to start a discussion (if they haven't already) on the
[mailing list](http://sourceforge.net/mailarchive/forum.php?forum_name=Kredit-development).

The patch will be accepted if there is broad consensus that it is a good thing.
Developers should expect to rework and resubmit patches if the code doesn't
match the project's coding conventions (see `doc/coding.md`) or are
controversial.

The `master` branch is regularly built and tested, but is not guaranteed to be
completely stable. [Tags](https://github.com/Kredit/Kredit/tags) are created
regularly to indicate new official, stable release versions of Kredit.

Testing
-------

Testing and code review is the bottleneck for development; we get more pull
requests than we can review and test. Please be patient and help out, and
remember this is a security-critical project where any mistake might cost people
lots of money.

### Automated Testing

Developers are strongly encouraged to write unit tests for new code, and to
submit new unit tests for old code.

Unit tests for the core code are in `src/test/`. To compile and run them:

    cd src; make -f makefile.unix test

Unit tests for the GUI code are in `src/qt/test/`. To compile and run them:

    qmake Kredit_QT_TEST=1 -o Makefile.test Kredit-qt.pro
    make -f Makefile.test
    ./Kredit-qt_test

Every pull request is built for both Windows and Linux on a dedicated server,
and unit and sanity tests are automatically run. The binaries produced may be
used for manual QA testing — a link to them will appear in a comment on the
pull request posted by [KreditPullTester](https://github.com/KreditPullTester). See https://github.com/TheBlueMatt/test-scripts
for the build/test scripts.

### Manual Quality Assurance (QA) Testing

Large changes should have a test plan, and should be tested by somebody other
than the developer who wrote the code.

See https://github.com/Kredit/QA/ for how to create a test plan.


debian/changelog
Kredit (0.8.1-natty3) natty; urgency=low

  * New pixmaps

 -- Jonas Schnelli <jonas.schnelli@include7.ch>  Mon, 13 May 2013 16:14:00 +0100

Kredit (0.8.1-natty2) natty; urgency=low

  * Remove dumb broken launcher script

 -- Matt Corallo <matt@bluematt.me>  Sun, 24 Mar 2013 20:01:00 -0400

Kredit (0.8.1-natty1) natty; urgency=low

  * New upstream release.

 -- Matt Corallo <matt@bluematt.me>  Tue, 19 Mar 2013 13:03:00 -0400

Kredit (0.8.0-natty1) natty; urgency=low

  * New upstream release.

 -- Matt Corallo <matt@bluematt.me>  Sat, 23 Feb 2013 16:01:00 -0500

Kredit (0.7.2-natty1) natty; urgency=low

  * New upstream release.

 -- Matt Corallo <matt@bluematt.me>  Sat, 15 Dec 2012 10:59:00 -0400

Kredit (0.7.1-natty1) natty; urgency=low

  * New upstream release.

 -- Matt Corallo <matt@bluematt.me>  Wed, 24 Oct 2012 15:06:00 -0400

Kredit (0.7.0-natty1) natty; urgency=low

  * New upstream release.

 -- Matt Corallo <matt@bluematt.me>  Mon, 17 Sep 2012 13:45:00 +0200

Kredit (0.6.3-natty1) natty; urgency=low

  * New upstream release.

 -- Matt Corallo <matt@bluematt.me>  Mon, 25 Jun 2012 23:47:00 +0200

Kredit (0.6.2-natty1) natty; urgency=low

  * Update package description and launch scripts.

 -- Matt Corallo <matt@bluematt.me>  Sat, 2 Jun 2012 16:41:00 +0200

Kredit (0.6.2-natty0) natty; urgency=low

  * New upstream release.

 -- Matt Corallo <matt@bluematt.me>  Tue, 8 May 2012 16:27:00 -0500

Kredit (0.6.1-natty0) natty; urgency=low

  * New upstream release.

 -- Matt Corallo <matt@bluematt.me>  Sun, 6 May 2012 20:09:00 -0500

Kredit (0.6.0-natty0) natty; urgency=low

  * New upstream release.
  * Add GNOME/KDE support for Kredit-qt's Kredit: URI support.
    Thanks to luke-jr for the KDE .protocol file.

 -- Matt Corallo <matt@bluematt.me>  Sat, 31 Mar 2012 15:35:00 -0500

Kredit (0.5.3-natty1) natty; urgency=low

  * Mark for upload to PPA.

 -- Matt Corallo <matt@bluematt.me>  Wed, 14 Mar 2012 23:06:00 -0400

Kredit (0.5.3-natty0) natty; urgency=low

  * New upstream release.

 -- Luke Dashjr <luke+Kredit+deb@dashjr.org>  Tue, 10 Jan 2012 15:57:00 -0500

Kredit (0.5.2-natty1) natty; urgency=low

  * Remove mentions on anonymity in package descriptions and manpage.
    These should never have been there, Kredit isnt anonymous without
    a ton of work that virtually no users will ever be willing and
    capable of doing

 -- Matt Corallo <matt@bluematt.me>  Sat, 7 Jan 2012 13:37:00 -0500

Kredit (0.5.2-natty0) natty; urgency=low

  * New upstream release.

 -- Luke Dashjr <luke+Kredit+deb@dashjr.org>  Fri, 16 Dec 2011 17:57:00 -0500

Kredit (0.5.1-natty0) natty; urgency=low

  * New upstream release.

 -- Matt Corallo <matt@bluematt.me>  Fri, 16 Dec 2011 13:27:00 -0500

Kredit (0.5.0-natty0) natty; urgency=low

  * New upstream release.

 -- Matt Corallo <matt@bluematt.me>  Mon, 21 Nov 2011 11:32:00 -0500

Kredit (0.5.0~rc7-natty0) natty; urgency=low

  * New upstream release candidate.

 -- Matt Corallo <matt@bluematt.me>  Sun, 20 Nov 2011 17:08:00 -0500

Kredit (0.5.0~rc3-natty0) natty; urgency=low

  * New upstream release candidate.
  * Don't set rpcpassword for Kredit-qt.

 -- Matt Corallo <matt@bluematt.me>  Tue, 8 Nov 2011 11:56:00 -0400

Kredit (0.5.0~rc1-natty1) natty; urgency=low

  * Add test_Kredit to build test
  * Fix clean
  * Remove uneccessary build-dependancies

 -- Matt Corallo <matt@bluematt.me>  Wed, 26 Oct 2011 14:37:18 -0400

Kredit (0.5.0~rc1-natty0) natty; urgency=low

  * Mark for natty
  * Fix broken build
  * Fix copyright listing
  * Remove Kredit: URL handler until Kredit actually has support for it (Oops)

 -- Matt Corallo <matt@bluematt.me>  Wed, 26 Oct 2011 14:37:18 -0400

Kredit (0.5.0~rc1-2) experimental; urgency=low

  * Add Kredit-qt

 -- Matt Corallo <matt@bluematt.me>  Tue, 25 Oct 2011 15:24:18 -0400

Kredit (0.5.0~rc1-1) experimental; urgency=low

  * New upstream prerelease.
  * Add Github as alternate upstream source in watch file.
  * Stop build-depending on libcrypto++-dev, and drop patch 1000:
    Upstream no longer use crypto++.
  * Drop patch 1003: Upstream builds dynamic by default now.
  * Update copyright file: Drop notes on longer included sources.

 -- Jonas Smedegaard <dr@jones.dk>  Fri, 14 Oct 2011 00:16:18 +0200

Kredit (0.4.0-1) unstable; urgency=low

  * New upstream release.
  * Stop repackaging source tarballs: No DFSG-violating stripping left.
  * Update copyright file:
    + Add Github URL to Source.
  * Drop dpkg-source local-options hint: Declared options are default
    since dpkg-source 1.16.1.
    + Add irc URL to Upstream-Contact.
    + Add comment on Kredit Developers to catch-all Files section.
    + Add Files sections for newly readded src/cryptopp/* (new custom
      BSD-like license), and newly added doc/build-osx.txt and
      src/makefile.osx (Expat).
  * Bump debhelper compatibility level to 7.
  * Suppress binary icns and gpg files.
  * Enable regression tests:
    + Build-depend on libboost-test-dev.
    + Extend patch 1003 to also dynamically link test binary.
    + Build and invoke test binary unless tests are disabled.
  * Tighten build-dependency on cdbs: Recent version needed to support
    debhelper 7.
  * Relax build-depend unversioned on debhelper: needed version
    satisfied even in oldstable.
  * Stop suppress optional build-dependencies: Satisfied in stable.
    Build-depend on devscripts (enabling copyright-check).

 -- Jonas Smedegaard <dr@jones.dk>  Wed, 05 Oct 2011 01:48:53 +0200

Kredit (0.3.24~dfsg-1) unstable; urgency=low

  * New upstream release.

  [ Jonas Smedegaard ]
  * Improve various usage hints:
    + Explicitly mention in long description that Kreditd contains
      daemon and command-line interface.
    + Extend README.Debian with section on lack of GUI, and add primary
      headline.
    + Avoid installing upstream README: contains no parts relevant for
      Debian usage.
    Thanks to richard for suggestions (see bug#629443).

tidy_datadir.sh
#!/bin/bash

if [ -d "$1" ]; then
  cd "$1"
else
  echo "Usage: $0 <datadir>" >&2
  echo "Removes obsolete Kredit database files" >&2
  exit 1
fi

LEVEL=0
if [ -f wallet.dat -a -f addr.dat -a -f blkindex.dat -a -f blk0001.dat ]; then LEVEL=1; fi
if [ -f wallet.dat -a -f peers.dat -a -f blkindex.dat -a -f blk0001.dat ]; then LEVEL=2; fi
if [ -f wallet.dat -a -f peers.dat -a -f coins/CURRENT -a -f blktree/CURRENT -a -f blocks/blk00000.dat ]; then LEVEL=3; fi
if [ -f wallet.dat -a -f peers.dat -a -f chainstate/CURRENT -a -f blocks/index/CURRENT -a -f blocks/blk00000.dat ]; then LEVEL=4; fi

case $LEVEL in
  0)
    echo "Error: no Kredit datadir detected."
    exit 1
    ;;
  1)
    echo "Detected old Kredit datadir (before 0.7)."
    echo "Nothing to do."
    exit 0
    ;;
  2)
    echo "Detected Kredit 0.7 datadir."
    ;;
  3)
    echo "Detected Kredit pre-0.8 datadir."
    ;;
  4)
    echo "Detected Kredit 0.8 datadir."
    ;;
esac

FILES=""
DIRS=""

if [ $LEVEL -ge 3 ]; then FILES=$(echo $FILES blk????.dat blkindex.dat); fi
if [ $LEVEL -ge 2 ]; then FILES=$(echo $FILES addr.dat); fi
if [ $LEVEL -ge 4 ]; then DIRS=$(echo $DIRS coins blktree); fi

for FILE in $FILES; do
  if [ -f $FILE ]; then
    echo "Deleting: $FILE"
    rm -f $FILE
  fi
done

for DIR in $DIRS; do
  if [ -d $DIR ]; then
    echo "Deleting: $DIR/"
    rm -rf $DIR
  fi
done

echo "Done."


bitrpc/bitrpc.py
from jsonrpc import ServiceProxy
import sys
import string

# ===== BEGIN USER SETTINGS =====
# if you do not set these you will be prompted for a password for every command
rpcuser = ""
rpcpass = ""
# ====== END USER SETTINGS ======


if rpcpass == "":
	access = ServiceProxy("http://127.0.0.1:8332")
else:
	access = ServiceProxy("http://"+rpcuser+":"+rpcpass+"@127.0.0.1:8332")
cmd = sys.argv[1].lower()

if cmd == "backupwallet":
	try:
		path = raw_input("Enter destination path/filename: ")
		print access.backupwallet(path)
	except:
		print "\n---An error occurred---\n"

elif cmd == "getaccount":
	try:
		addr = raw_input("Enter a Kredit address: ")
		print access.getaccount(addr)
	except:
		print "\n---An error occurred---\n"

elif cmd == "getaccountaddress":
	try:
		acct = raw_input("Enter an account name: ")
		print access.getaccountaddress(acct)
	except:
		print "\n---An error occurred---\n"

elif cmd == "getaddressesbyaccount":
	try:
		acct = raw_input("Enter an account name: ")
		print access.getaddressesbyaccount(acct)
	except:
		print "\n---An error occurred---\n"

elif cmd == "getbalance":
	try:
		acct = raw_input("Enter an account (optional): ")
		mc = raw_input("Minimum confirmations (optional): ")
		try:
			print access.getbalance(acct, mc)
		except:
			print access.getbalance()
	except:
		print "\n---An error occurred---\n"

elif cmd == "getblockbycount":
	try:
		height = raw_input("Height: ")
		print access.getblockbycount(height)
	except:
		print "\n---An error occurred---\n"

elif cmd == "getblockcount":
	try:
		print access.getblockcount()
	except:
		print "\n---An error occurred---\n"

elif cmd == "getblocknumber":
	try:
		print access.getblocknumber()
	except:
		print "\n---An error occurred---\n"

elif cmd == "getconnectioncount":
	try:
		print access.getconnectioncount()
	except:
		print "\n---An error occurred---\n"

elif cmd == "getdifficulty":
	try:
		print access.getdifficulty()
	except:
		print "\n---An error occurred---\n"

elif cmd == "getgenerate":
	try:
		print access.getgenerate()
	except:
		print "\n---An error occurred---\n"

elif cmd == "gethashespersec":
	try:
		print access.gethashespersec()
	except:
		print "\n---An error occurred---\n"

elif cmd == "getinfo":
	try:
		print access.getinfo()
	except:
		print "\n---An error occurred---\n"

elif cmd == "getnewaddress":
	try:
		acct = raw_input("Enter an account name: ")
		try:
			print access.getnewaddress(acct)
		except:
			print access.getnewaddress()
	except:
		print "\n---An error occurred---\n"

elif cmd == "getreceivedbyaccount":
	try:
		acct = raw_input("Enter an account (optional): ")
		mc = raw_input("Minimum confirmations (optional): ")
		try:
			print access.getreceivedbyaccount(acct, mc)
		except:
			print access.getreceivedbyaccount()
	except:
		print "\n---An error occurred---\n"

elif cmd == "getreceivedbyaddress":
	try:
		addr = raw_input("Enter a Kredit address (optional): ")
		mc = raw_input("Minimum confirmations (optional): ")
		try:
			print access.getreceivedbyaddress(addr, mc)
		except:
			print access.getreceivedbyaddress()
	except:
		print "\n---An error occurred---\n"

elif cmd == "gettransaction":
	try:
		txid = raw_input("Enter a transaction ID: ")
		print access.gettransaction(txid)
	except:
		print "\n---An error occurred---\n"

elif cmd == "getwork":
	try:
		data = raw_input("Data (optional): ")
		try:
			print access.gettransaction(data)
		except:
			print access.gettransaction()
	except:
		print "\n---An error occurred---\n"

elif cmd == "help":
	try:
		cmd = raw_input("Command (optional): ")
		try:
			print access.help(cmd)
		except:
			print access.help()
	except:
		print "\n---An error occurred---\n"

elif cmd == "listaccounts":
	try:
		mc = raw_input("Minimum confirmations (optional): ")
		try:
			print access.listaccounts(mc)
		except:
			print access.listaccounts()
	except:
		print "\n---An error occurred---\n"

elif cmd == "listreceivedbyaccount":
	try:
		mc = raw_input("Minimum confirmations (optional): ")
		incemp = raw_input("Include empty? (true/false, optional): ")
		try:
			print access.listreceivedbyaccount(mc, incemp)
		except:
			print access.listreceivedbyaccount()
	except:
		print "\n---An error occurred---\n"

elif cmd == "listreceivedbyaddress":
	try:
		mc = raw_input("Minimum confirmations (optional): ")
		incemp = raw_input("Include empty? (true/false, optional): ")
		try:
			print access.listreceivedbyaddress(mc, incemp)
		except:
			print access.listreceivedbyaddress()
	except:
		print "\n---An error occurred---\n"

elif cmd == "listtransactions":
	try:
		acct = raw_input("Account (optional): ")
		count = raw_input("Number of transactions (optional): ")
		frm = raw_input("Skip (optional):")
		try:
			print access.listtransactions(acct, count, frm)
		except:
			print access.listtransactions()
	except:
		print "\n---An error occurred---\n"

elif cmd == "move":
	try:
		frm = raw_input("From: ")
		to = raw_input("To: ")
		amt = raw_input("Amount:")
		mc = raw_input("Minimum confirmations (optional): ")
		comment = raw_input("Comment (optional): ")
		try:
			print access.move(frm, to, amt, mc, comment)
		except:
			print access.move(frm, to, amt)
	except:
		print "\n---An error occurred---\n"

elif cmd == "sendfrom":
	try:
		frm = raw_input("From: ")
		to = raw_input("To: ")
		amt = raw_input("Amount:")
		mc = raw_input("Minimum confirmations (optional): ")
		comment = raw_input("Comment (optional): ")
		commentto = raw_input("Comment-to (optional): ")
		try:
			print access.sendfrom(frm, to, amt, mc, comment, commentto)
		except:
			print access.sendfrom(frm, to, amt)
	except:
		print "\n---An error occurred---\n"

elif cmd == "sendmany":
	try:
		frm = raw_input("From: ")
		to = raw_input("To (in format address1:amount1,address2:amount2,...): ")
		mc = raw_input("Minimum confirmations (optional): ")
		comment = raw_input("Comment (optional): ")
		try:
			print access.sendmany(frm,to,mc,comment)
		except:
			print access.sendmany(frm,to)
	except:
		print "\n---An error occurred---\n"

elif cmd == "sendtoaddress":
	try:
		to = raw_input("To (in format address1:amount1,address2:amount2,...): ")
		amt = raw_input("Amount:")
		comment = raw_input("Comment (optional): ")
		commentto = raw_input("Comment-to (optional): ")
		try:
			print access.sendtoaddress(to,amt,comment,commentto)
		except:
			print access.sendtoaddress(to,amt)
	except:
		print "\n---An error occurred---\n"

elif cmd == "setaccount":
	try:
		addr = raw_input("Address: ")
		acct = raw_input("Account:")
		print access.setaccount(addr,acct)
	except:
		print "\n---An error occurred---\n"

elif cmd == "setgenerate":
	try:
		gen= raw_input("Generate? (true/false): ")
		cpus = raw_input("Max processors/cores (-1 for unlimited, optional):")
		try:
			print access.setgenerate(gen, cpus)
		except:
			print access.setgenerate(gen)
	except:
		print "\n---An error occurred---\n"

elif cmd == "settxfee":
	try:
		amt = raw_input("Amount:")
		print access.settxfee(amt)
	except:
		print "\n---An error occurred---\n"

elif cmd == "stop":
	try:
		print access.stop()
	except:
		print "\n---An error occurred---\n"

elif cmd == "validateaddress":
	try:
		addr = raw_input("Address: ")
		print access.validateaddress(addr)
	except:
		print "\n---An error occurred---\n"

elif cmd == "walletpassphrase":
	try:
		pwd = raw_input("Enter wallet passphrase: ")
		access.walletpassphrase(pwd, 60)
		print "\n---Wallet unlocked---\n"
	except:
		print "\n---An error occurred---\n"

elif cmd == "walletpassphrasechange":
	try:
		pwd = raw_input("Enter old wallet passphrase: ")
		pwd2 = raw_input("Enter new wallet passphrase: ")
		access.walletpassphrasechange(pwd, pwd2)
		print
		print "\n---Passphrase changed---\n"
	except:
		print
		print "\n---An error occurred---\n"
		print

else:
	print "Command not found or not supported"

testgen/base58.py
'''
Kredit base58 encoding and decoding.

Based on https://Kredittalk.org/index.php?topic=1026.0 (public domain)
'''
import hashlib

# for compatibility with following code...
class SHA256:
    new = hashlib.sha256

if str != bytes:
    # Python 3.x
    def ord(c):
        return c
    def chr(n):
        return bytes( (n,) )

__b58chars = '123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz'
__b58base = len(__b58chars)
b58chars = __b58chars

def b58encode(v):
    """ encode v, which is a string of bytes, to base58.
    """
    long_value = 0
    for (i, c) in enumerate(v[::-1]):
        long_value += (256**i) * ord(c)

    result = ''
    while long_value >= __b58base:
        div, mod = divmod(long_value, __b58base)
        result = __b58chars[mod] + result
        long_value = div
    result = __b58chars[long_value] + result

    # Kredit does a little leading-zero-compression:
    # leading 0-bytes in the input become leading-1s
    nPad = 0
    for c in v:
        if c == '\0': nPad += 1
        else: break

    return (__b58chars[0]*nPad) + result

def b58decode(v, length = None):
    """ decode v into a string of len bytes
    """
    long_value = 0
    for (i, c) in enumerate(v[::-1]):
        long_value += __b58chars.find(c) * (__b58base**i)

    result = bytes()
    while long_value >= 256:
        div, mod = divmod(long_value, 256)
        result = chr(mod) + result
        long_value = div
    result = chr(long_value) + result

    nPad = 0
    for c in v:
        if c == __b58chars[0]: nPad += 1
        else: break

    result = chr(0)*nPad + result
    if length is not None and len(result) != length:
        return None

    return result

def checksum(v):
    """Return 32-bit checksum based on SHA256"""
    return SHA256.new(SHA256.new(v).digest()).digest()[0:4]

def b58encode_chk(v):
    """b58encode a string, with 32-bit checksum"""
    return b58encode(v + checksum(v))

def b58decode_chk(v):
    """decode a base58 string, check and remove checksum"""
    result = b58decode(v)
    if result is None:
        return None
    h3 = checksum(result[:-4])
    if result[-4:] == checksum(result[:-4]):
        return result[:-4]
    else:
        return None

def get_bcaddress_version(strAddress):
    """ Returns None if strAddress is invalid.  Otherwise returns integer version of address. """
    addr = b58decode_chk(strAddress)
    if addr is None or len(addr)!=21: return None
    version = addr[0]
    return ord(version)

if __name__ == '__main__':
    # Test case (from http://gitorious.org/Kredit/python-base58.git)
    assert get_bcaddress_version('15VjRaDX9zpbA8LVnbrCAFzrVzN7ixHNsC') is 0
    _ohai = 'o hai'.encode('ascii')
    _tmp = b58encode(_ohai)
    assert _tmp == 'DYB3oMS'
    assert b58decode(_tmp, 5) == _ohai
    print("Tests passed")


verifysfbinaries/verify.sh
#!/bin/bash

###   This script attempts to download the signature file SHA256SUMS.asc from SourceForge
###   It first checks if the signature passes, and then downloads the files specified in
###   the file, and checks if the hashes of these files match those that are specified
###   in the signature file.
###   The script returns 0 if everything passes the checks. It returns 1 if either the
###   signature check or the hash check doesn't pass. If an error occurs the return value is 2

function clean_up {
   for file in $*
   do
      rm "$file" 2> /dev/null
   done
}

WORKINGDIR="/tmp/Kredit"
TMPFILE="hashes.tmp"

#this URL is used if a version number is not specified as an argument to the script
SIGNATUREFILE="http://downloads.sourceforge.net/project/Kredit/Kredit/Kredit-0.8.6/SHA256SUMS.asc"

SIGNATUREFILENAME="SHA256SUMS.asc"
RCSUBDIR="test/"
BASEDIR="http://downloads.sourceforge.net/project/Kredit/Kredit/"
VERSIONPREFIX="Kredit-"
RCVERSIONSTRING="rc"

if [ ! -d "$WORKINGDIR" ]; then
   mkdir "$WORKINGDIR"
fi

cd "$WORKINGDIR"

#test if a version number has been passed as an argument
if [ -n "$1" ]; then
   #let's also check if the version number includes the prefix 'Kredit-',
   #  and add this prefix if it doesn't
   if [[ $1 == "$VERSIONPREFIX"* ]]; then
      VERSION="$1"
   else
      VERSION="$VERSIONPREFIX$1"
   fi

   #now let's see if the version string contains "rc", and strip it off if it does
   #  and simultaneously add RCSUBDIR to BASEDIR, where we will look for SIGNATUREFILENAME
   if [[ $VERSION == *"$RCVERSIONSTRING"* ]]; then
      BASEDIR="$BASEDIR${VERSION/%-$RCVERSIONSTRING*}/"
      BASEDIR="$BASEDIR$RCSUBDIR"
   else
      BASEDIR="$BASEDIR$VERSION/"
   fi

   SIGNATUREFILE="$BASEDIR$SIGNATUREFILENAME"
else
   BASEDIR="${SIGNATUREFILE%/*}/"
fi

#first we fetch the file containing the signature
WGETOUT=$(wget -N "$BASEDIR$SIGNATUREFILENAME" 2>&1)

#and then see if wget completed successfully
if [ $? -ne 0 ]; then
   echo "Error: couldn't fetch signature file. Have you specified the version number in the following format?"
   echo "[Kredit-]<version>-[rc[0-9]] (example: Kredit-0.7.1-rc1)"
   echo "wget output:"
   echo "$WGETOUT"|sed 's/^/\t/g'
   exit 2
fi

#then we check it
GPGOUT=$(gpg --yes --decrypt --output "$TMPFILE" "$SIGNATUREFILENAME" 2>&1)

#return value 0: good signature
#return value 1: bad signature
#return value 2: gpg error

RET="$?"
if [ $RET -ne 0 ]; then
   if [ $RET -eq 1 ]; then
      #and notify the user if it's bad
      echo "Bad signature."
   elif [ $RET -eq 2 ]; then
      #or if a gpg error has occurred
      echo "gpg error. Do you have Gavin's code signing key installed?"
   fi

   echo "gpg output:"
   echo "$GPGOUT"|sed 's/^/\t/g'
   clean_up $SIGNATUREFILENAME $TMPFILE
   exit "$RET"
fi

#here we extract the filenames from the signature file
FILES=$(awk '{print $2}' "$TMPFILE")

#and download these one by one
for file in in $FILES
do
   wget --quiet -N "$BASEDIR$file"
done

#check hashes
DIFF=$(diff <(sha256sum $FILES) "$TMPFILE")

if [ $? -eq 1 ]; then
   echo "Hashes don't match."
   echo "Offending files:"
   echo "$DIFF"|grep "^<"|awk '{print "\t"$3}'
   exit 1
elif [ $? -gt 1 ]; then
   echo "Error executing 'diff'"
   exit 2   
fi

#everything matches! clean up the mess
clean_up $FILES $SIGNATUREFILENAME $TMPFILE

exit 0


pyminer/pyminer.py
#!/usr/bin/python
#
# Copyright (c) 2011 The Kredit developers
# Distributed under the MIT/X11 software license, see the accompanying
# file COPYING or http://www.opensource.org/licenses/mit-license.php.
#

import time
import json
import pprint
import hashlib
import struct
import re
import base64
import httplib
import sys
from multiprocessing import Process

ERR_SLEEP = 15
MAX_NONCE = 1000000L

settings = {}
pp = pprint.PrettyPrinter(indent=4)

class KreditRPC:
	OBJID = 1

	def __init__(self, host, port, username, password):
		authpair = "%s:%s" % (username, password)
		self.authhdr = "Basic %s" % (base64.b64encode(authpair))
		self.conn = httplib.HTTPConnection(host, port, False, 30)
	def rpc(self, method, params=None):
		self.OBJID += 1
		obj = { 'version' : '1.1',
			'method' : method,
			'id' : self.OBJID }
		if params is None:
			obj['params'] = []
		else:
			obj['params'] = params
		self.conn.request('POST', '/', json.dumps(obj),
			{ 'Authorization' : self.authhdr,
			  'Content-type' : 'application/json' })

		resp = self.conn.getresponse()
		if resp is None:
			print "JSON-RPC: no response"
			return None

		body = resp.read()
		resp_obj = json.loads(body)
		if resp_obj is None:
			print "JSON-RPC: cannot JSON-decode body"
			return None
		if 'error' in resp_obj and resp_obj['error'] != None:
			return resp_obj['error']
		if 'result' not in resp_obj:
			print "JSON-RPC: no result in object"
			return None

		return resp_obj['result']
	def getblockcount(self):
		return self.rpc('getblockcount')
	def getwork(self, data=None):
		return self.rpc('getwork', data)

def uint32(x):
	return x & 0xffffffffL

def bytereverse(x):
	return uint32(( ((x) << 24) | (((x) << 8) & 0x00ff0000) |
			(((x) >> 8) & 0x0000ff00) | ((x) >> 24) ))

def bufreverse(in_buf):
	out_words = []
	for i in range(0, len(in_buf), 4):
		word = struct.unpack('@I', in_buf[i:i+4])[0]
		out_words.append(struct.pack('@I', bytereverse(word)))
	return ''.join(out_words)

def wordreverse(in_buf):
	out_words = []
	for i in range(0, len(in_buf), 4):
		out_words.append(in_buf[i:i+4])
	out_words.reverse()
	return ''.join(out_words)

class Miner:
	def __init__(self, id):
		self.id = id
		self.max_nonce = MAX_NONCE

	def work(self, datastr, targetstr):
		# decode work data hex string to binary
		static_data = datastr.decode('hex')
		static_data = bufreverse(static_data)

		# the first 76b of 80b do not change
		blk_hdr = static_data[:76]

		# decode 256-bit target value
		targetbin = targetstr.decode('hex')
		targetbin = targetbin[::-1]	# byte-swap and dword-swap
		targetbin_str = targetbin.encode('hex')
		target = long(targetbin_str, 16)

		# pre-hash first 76b of block header
		static_hash = hashlib.sha256()
		static_hash.update(blk_hdr)

		for nonce in xrange(self.max_nonce):

			# encode 32-bit nonce value
			nonce_bin = struct.pack("<I", nonce)

			# hash final 4b, the nonce value
			hash1_o = static_hash.copy()
			hash1_o.update(nonce_bin)
			hash1 = hash1_o.digest()

			# sha256 hash of sha256 hash
			hash_o = hashlib.sha256()
			hash_o.update(hash1)
			hash = hash_o.digest()

			# quick test for winning solution: high 32 bits zero?
			if hash[-4:] != '\0\0\0\0':
				continue

			# convert binary hash to 256-bit Python long
			hash = bufreverse(hash)
			hash = wordreverse(hash)

			hash_str = hash.encode('hex')
			l = long(hash_str, 16)

			# proof-of-work test:  hash < target
			if l < target:
				print time.asctime(), "PROOF-OF-WORK found: %064x" % (l,)
				return (nonce + 1, nonce_bin)
			else:
				print time.asctime(), "PROOF-OF-WORK false positive %064x" % (l,)
#				return (nonce + 1, nonce_bin)

		return (nonce + 1, None)

	def submit_work(self, rpc, original_data, nonce_bin):
		nonce_bin = bufreverse(nonce_bin)
		nonce = nonce_bin.encode('hex')
		solution = original_data[:152] + nonce + original_data[160:256]
		param_arr = [ solution ]
		result = rpc.getwork(param_arr)
		print time.asctime(), "--> Upstream RPC result:", result

	def iterate(self, rpc):
		work = rpc.getwork()
		if work is None:
			time.sleep(ERR_SLEEP)
			return
		if 'data' not in work or 'target' not in work:
			time.sleep(ERR_SLEEP)
			return

		time_start = time.time()

		(hashes_done, nonce_bin) = self.work(work['data'],
						     work['target'])

		time_end = time.time()
		time_diff = time_end - time_start

		self.max_nonce = long(
			(hashes_done * settings['scantime']) / time_diff)
		if self.max_nonce > 0xfffffffaL:
			self.max_nonce = 0xfffffffaL

		if settings['hashmeter']:
			print "HashMeter(%d): %d hashes, %.2f Khash/sec" % (
			      self.id, hashes_done,
			      (hashes_done / 1000.0) / time_diff)

		if nonce_bin is not None:
			self.submit_work(rpc, work['data'], nonce_bin)

	def loop(self):
		rpc = KreditRPC(settings['host'], settings['port'],
				 settings['rpcuser'], settings['rpcpass'])
		if rpc is None:
			return

		while True:
			self.iterate(rpc)

def miner_thread(id):
	miner = Miner(id)
	miner.loop()

if __name__ == '__main__':
	if len(sys.argv) != 2:
		print "Usage: pyminer.py CONFIG-FILE"
		sys.exit(1)

	f = open(sys.argv[1])
	for line in f:
		# skip comment lines
		m = re.search('^\s*#', line)
		if m:
			continue

		# parse key=value lines
		m = re.search('^(\w+)\s*=\s*(\S.*)$', line)
		if m is None:
			continue
		settings[m.group(1)] = m.group(2)
	f.close()

	if 'host' not in settings:
		settings['host'] = '127.0.0.1'
	if 'port' not in settings:
		settings['port'] = 8332
	if 'threads' not in settings:
		settings['threads'] = 1
	if 'hashmeter' not in settings:
		settings['hashmeter'] = 0
	if 'scantime' not in settings:
		settings['scantime'] = 30L
	if 'rpcuser' not in settings or 'rpcpass' not in settings:
		print "Missing username and/or password in cfg file"
		sys.exit(1)

	settings['port'] = int(settings['port'])
	settings['threads'] = int(settings['threads'])
	settings['hashmeter'] = int(settings['hashmeter'])
	settings['scantime'] = long(settings['scantime'])

	thr_list = []
	for thr_id in range(settings['threads']):
		p = Process(target=miner_thread, args=(thr_id,))
		p.start()
		thr_list.append(p)
		time.sleep(1)			# stagger threads

	print settings['threads'], "mining threads started"

	print time.asctime(), "Miner Starts - %s:%s" % (settings['host'], settings['port'])
	try:
		for thr_proc in thr_list:
			thr_proc.join()
	except KeyboardInterrupt:
		pass
	print time.asctime(), "Miner Stops - %s:%s" % (settings['host'], settings['port'])

